# Прикладной анализ данных в социальных науках

## 02. Дескриптивный анализ данных

Меньше и больше счастья

Всемирный доклад ООН о счастье содержит данные об опросе людей в разных странах, в котором они оценивают свой общий уровень счастья, а также уровне удовлетворения в отдельных аспектах. В рамках сбора данных для репрезентативных национальных выборок респондентов просят ответить на ряд вопросов по шкале от нуля до десяти, где десять — это лучшая из возможных жизней, а ноль — худшая. Ответы отдельных респондентов усредняются для получения общего для страны показателя. Попробуем исследовать данные за 2019 год, чтобы лучше понять их структуру. CSV-файл доступен под названием “2019.csv”.

Некоторые переменные, доступные в данных:

    'Country or region' - название страны
    'Score' - Уровень индекса счастья
    'GDP per capita' - Уровень ВВП на душу населения

Какое максимальное и минимальное значение индекса счастья в 2019 году?
Формат ввода

Важно!

Два ответа должны быть на разных строчках, это можно сделать, используя print два раза, например:

print('значение максимума')

print('значение минимума')
Формат вывода

6.232
3.243

In [1]:
import pandas as pd

In [4]:
df = pd.read_csv('2019.csv')

In [7]:
print(df.Score.max())
print(df.Score.min())

7.769
2.853


Средний уровень счастья

Всемирный доклад ООН о счастье содержит данные об опросе людей в разных странах, в котором они оценивают свой общий уровень счастья, а также уровне удовлетворения в отдельных аспектах. В рамках сбора данных для репрезентативных национальных выборок респондентов просят ответить на ряд вопросов по шкале от нуля до десяти, где десять — это лучшая из возможных жизней, а ноль — худшая. Ответы отдельных респондентов усредняются для получения общего для страны показателя. Попробуем исследовать данные за 2019 год, чтобы лучше понять их структуру. CSV-файл доступен под названием “2019.csv”.

Какой средний уровень счастья среди всех стран?

Некоторые переменные, доступные в данных:

    'Country or region' - название страны
    'Score' - Уровень индекса счастья
    'GDP per capita' - Уровень ВВП на душу населения

Какой средний уровень счастья среди всех стран?
Формат вывода

6.232

In [8]:
df.head()

Overall rank Country or region  Score  GDP per capita  Social support  \
0             1           Finland  7.769           1.340           1.587   
1             2           Denmark  7.600           1.383           1.573   
2             3            Norway  7.554           1.488           1.582   
3             4           Iceland  7.494           1.380           1.624   
4             5       Netherlands  7.488           1.396           1.522   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
0                    0.986                         0.596       0.153   
1                    0.996                         0.592       0.252   
2                    1.028                         0.603       0.271   
3                    1.026                         0.591       0.354   
4                    0.999                         0.557       0.322   

   Perceptions of corruption  
0                      0.393  
1                      0.410  
2                      0.341  
3                      0.118  
4                      0.298

In [10]:
print(df['Score'].mean())

5.407096153846155


Каков медианный уровень счастья среди всех стран?

In [11]:
print(df['Score'].median())

5.3795


Какова мода среди всех стран?

Учтите, что так как мод может быть несколько, то функция mode() возвращает массив, в котором может быть одно или несколько чисел. Мы точно знаем, что в этой задаче единственная мода. Чтобы вернуть её как число, добавьте явное указание, что вам нужен нулевой элемент из массива: df['col_name'].mode()[0].

In [13]:
print(df['Score'].mode()[0])

5.208


Каково стандартное отклонение среди всех стран?

In [14]:
print(df['Score'].std())

1.1131198687956712


Найдите десять самых счастливых стран.

Отсортируйте наблюдения по уровню счастья, затем выберите единственную колонку с названиями регионов и выведите первые 10 наблюдений. При выдаче значения должны быть выведены в качестве списка, используйте для этого функцию .tolist(). 

In [21]:
print(df.sort_values(by='Score', ascending=False)['Country or region'][0:10].tolist())

['Finland', 'Denmark', 'Norway', 'Iceland', 'Netherlands', 'Switzerland', 'Sweden', 'New Zealand', 'Canada', 'Austria']


Подсчитайте суммарный ВВП всех стран.

In [23]:
print(df['GDP per capita'].sum())

141.203


Подсчитайте суммарный ВВП первых десяти стран.

In [29]:
print(df[0:10]['GDP per capita'].sum())

13.87


## 03. Парные сравнения числовых данных

In [30]:
df = pd.read_csv('books_prep.csv')

In [31]:
df.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  \
0                  JJ Smith          4.7    17350      8  2016  Non Fiction   
1              Stephen King          4.6     2052     22  2011      Fiction   
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction   
3             George Orwell          4.7    21424      6  2017      Fiction   
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction   

   User Rating (Round) Price (Above Average)  
0                    5                    No  
1                    5                   Yes  
2                    5                   Yes  
3                    5                    No  
4                    5                    No

Самый популярный писатель

В этой и в следующих задачах мы используем набор данных с самыми популярными книгами на Amazon. Мы добавили в него несколько дополнительных переменных. CSV-файл доступен под названием “books_prep.csv”.

В прошлых главах мы говорили о том, что метод describe выводит несколько полезных описательных статистик. Мы получаем результат в виде серии (так называется датафрейм из одной колонки), поэтому мы можем вывести отдельный показатель, используя стандартный синтаксис pandas.

Например, чтобы вывести среднее значение, мы можем использовать следующую инструкцию: df['col_name'].describe()['mean']

Выведите имя самого популярного писателя в наборе данных

Чтобы решить задачу, используйте описанный выше синтаксис уточнений к методу describe.

In [83]:
print(df['Author'].describe()['top'])

Jeff Kinney


Определите средние значения рейтинга для «дешёвых» и «дорогих» книг.

Попробуем исследовать этот набор данных. Мы создали переменную “Price (Above Average)”, которая содержит “Yes”, если цена книги больше средней, и “No”, если меньше.

Мы предполагаем, что рейтинг книги не зависит от её стоимости. Давайте проверим эту гипотезу. Начнём с того, что выведем средние значения пользовательского рейтинга для каждой из групп.

Вам необходимо загрузить набор данных и напечатать средний рейтинг для «дорогих» книг. CSV-файл доступен под названием “books_prep.csv”.

Округлите ответ до 2 числа после запятой с помощью команды round.

In [91]:
exp_avg = df[df['Price (Above Average)']  == 'Yes']['User Rating'].mean().round(2)
print(exp_avg)

4.56


Определите средние значения рейтинга для «дешёвых» и «дорогих» книг.

А теперь вам необходимо загрузить набор данных и напечатать средний рейтинг для «дешёвых» книг. CSV-файл доступен под названием “books_prep.csv”.

Округлите ответ до 2 числа после запятой с помощью команды round.

In [92]:
chp_avg = df[df['Price (Above Average)']  == 'No']['User Rating'].mean().round(2)
print(chp_avg)

4.65


Проведите тест для определения равенства дисперсий. Выведите значение p-value.

Прежде чем сравнивать средние, нам необходимо сравнить дисперсии. Для этого мы можем использовать критерий Левене.

Проведите тест для определения равенства дисперсий у “дешёвых” и “дорогих” книг. Выведите значение p-value.

Вам необходимо вывести только значение p-value и округлить его до 2 знаков после запятой с помощью команды round. Попробуйте немного изменить функцию levene('add something here').pvalue. CSV-файл доступен под названием “books_prep.csv”.

In [93]:
from scipy.stats import levene

In [104]:
exp_grp = df[df['Price (Above Average)']  == 'Yes']['User Rating']
chp_grp = df[df['Price (Above Average)']  == 'No']['User Rating']

p_val = levene(exp_grp, chp_grp).pvalue

print(round(p_val, 2))

0.1


Проведите тест для определения равенства средних. Выведите значение p-value

Теперь давайте попробуем понять, является ли значимым то различие между средними, которое мы увидели. Вам снова необходимо вывести только значение p-value и округлить его до 2 знаков после запятой с помощью команды round. Модифицировать метод необходимо аналогично предыдущей задаче.

CSV-файл доступен под названием “books_prep.csv”.

In [106]:
from scipy.stats import ttest_ind

In [110]:
p_val = ttest_ind(exp_grp, chp_grp, equal_var=True).pvalue

print(round(p_val, 2))

8.190934970309897e-06

ANOVA
Проведите тест ANOVA. Выведите значение p-value

Ещё в этом примере мы создали переменную “User Rating (Round)”, которая содержит округлённые до целого значения среднего пользовательского рейтинга. Мы предполагаем, что рейтинг книги для каждой из групп, — у нас получилось их три, — зависит от числа рецензий (на хорошие книги люди охотнее пишут рецензии). Давайте проверим эту гипотезу с помощью теста ANOVA.

Вам необходимо загрузить набор данных и вывести только значение p-value, округленное до 2 знаков после запятой. CSV-файл доступен под названием “books_prep.csv”.

In [123]:
from scipy.stats import f_oneway

In [127]:
grp_5 = df[df['User Rating (Round)'] == 5]['Reviews']
grp_4 = df[df['User Rating (Round)'] == 4]['Reviews']
grp_3 = df[df['User Rating (Round)'] == 3]['Reviews']

p_val = f_oneway(grp_5, grp_4, grp_3).pvalue

print(round(p_val, 2))

0.3


## 04. Категориальные данные

Ознакомление с датасетом

В этой и последующих задачах мы будем работать с датасетом влияния COVID-19 на образование, психическое здоровье и распорядок дня учащихся. Ознакомиться с полной версией датасета можно по ссылке: https://www.kaggle.com/datasets/kunal28chaturvedi/covid19-and-its-impact-on-students.

Датасет доступен под названием 'COVID-19 SSR.csv'

В датасете содержатся такие переменные как время на занятия, самостоятельное обучение, просмотр страниц в социальных сетях в день, а также соц-демографические характеристики учащихся (возраст, регион проживания) и другие.

Вам необходимо вывести наиболее частый тип переменных встречаемый в датасете.

In [2]:
file_path = 'COVID-19 SSR.csv'
df = pd.read_csv(file_path)

In [14]:
df.head()

Unnamed: 0  ID Region of residence  Age of Subject  \
0           0  R1           Delhi-NCR              21   
1           1  R2           Delhi-NCR              21   
2           2  R3           Delhi-NCR              20   
3           3  R4           Delhi-NCR              20   
4           4  R5           Delhi-NCR              21   

   Time spent on Online Class Rating of Online Class experience  \
0                         2.0                              GOOD   
1                         0.0                         EXCELLENT   
2                         7.0                         VERY POOR   
3                         3.0                         VERY POOR   
4                         3.0                              GOOD   

  Medium for online class  Time spent on self study  Time spent on fitness  \
0          Laptop/Desktop                       4.0                    0.0   
1              Smartphone                       0.0                    2.0   
2          Laptop/Desktop                       3.0                    0.0   
3              Smartphone                       2.0                    1.0   
4          Laptop/Desktop                       3.0                    1.0   

   Time spent on sleep  Time spent on social media  \
0                  7.0                         3.0   
1                 10.0                         3.0   
2                  6.0                         2.0   
3                  6.0                         5.0   
4                  8.0                         3.0   

  Prefered social media platform Time spent on TV  Number of meals per day  \
0                       Linkedin                1                        4   
1                        Youtube                0                        3   
2                       Linkedin                0                        3   
3                      Instagram                0                        3   
4                      Instagram                1                        4   

  Change in your weight Health issue during lockdown  \
0             Increased                           NO   
1             Decreased                           NO   
2       Remain Constant                           NO   
3             Decreased                           NO   
4       Remain Constant                           NO   

                   Stress busters Time utilized  \
0                         Cooking           YES   
1  Scrolling through social media           YES   
2              Listening to music            NO   
3             Watching web series            NO   
4                    Social Media            NO   

  Do you find yourself more connected with your family, close friends , relatives  ?  \
0                                                YES                                   
1                                                 NO                                   
2                                                YES                                   
3                                                 NO                                   
4                                                 NO                                   

  What you miss the most  
0         School/college  
1  Roaming around freely  
2             Travelling  
3    Friends , relatives  
4             Travelling

In [78]:
df2 = df.dtypes.value_counts().reset_index()
df2

index  count
0   object     12
1  float64      5
2    int64      3

In [89]:
print(df.dtypes.value_counts().idxmax())

object


Изменяем капслок

Не всегда данные в одной колонке имеют одинаковое оформление. Например, одно значение написано буквами верхнего регистра, а другое - нижнего. Для приведения значений в категориальных переменных к одному виду существует набор функций.

title() - преобразовывает слова в формате title, где первое слово в строке начинается с символа верхнего регистра, а остальные приводятся к символам нижнего регистра (Например, "привет, студенты!" преобразуется в "Привет, студенты!").

В датасете'COVID-19 SSR.csv' в переменной 'Rating of Online Class experience' все значения записаны символами верхнего регистра (Например, "GOOD").

Вам необходимо привести значения переменной 'Rating of Online Class experience' к значению "GOOD" с помощью функции title(). Например, df('Название переменной').str.title().

В выдаче выведите пять первых значения переменной 'Rating of Online Class experience'.

Датасет доступен под названием 'COVID-19 SSR.csv'
Формат вывода

0 Good
1 Very Poor
2 Good
3 Very Poor
4 Excellent
Name: Rating of Online Class experience, dtype: object

In [3]:
df['Rating of Online Class experience'] = df['Rating of Online Class experience'].str.title()
print(df['Rating of Online Class experience'].head(5))

0         Good
1    Excellent
2    Very Poor
3    Very Poor
4         Good
Name: Rating of Online Class experience, dtype: object


Сколько спят учащиеся?

В нашем датасете есть переменная, указывающая на среднее количество сна в день ['Time spent on sleep']. Для здорового человека нормально спать 7-8 часов. Но учащиеся могут спать больше или меньше необходимого времени сна. То есть они могут спать например меньше 7 (6.9 часов) и больше 8 часов. Оптимальным количеством сна является 7-8 часов сна.

Вам необходимо вывести количество учащихся, которые спят больше или меньше оптимального времени для сна. Одним из способом решения этой переменной будут следующие шаги:

    Воспользоваться циклом для создания новой переменной (название которой может быть 'Sleep') по переменной['Time spent on sleep'] , где необходимо прописать следующие условия. Если ['Time spent on sleep'] меньше 9, но при этом больше 6.9, то это нормальное количество сна ('normal'), в другом случае значение переменной будет 'not normal'.
    Сделать фильтр по новой переменной, таким образом, чтобы мы оставили строки со значением сна'not normal'.
    Вывести количество строк отфильтрованного датасета (например, с помощью функции 'len()').

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [105]:
df = pd.read_csv('COVID-19 Survey Student Responses.csv')
print(sum((df['Time spent on sleep'] < 7) | (df['Time spent on sleep'] > 8.9)))

529


Мы обратим внимание на переменную 'Time spent on TV', которая содержит в себе не только числовые показатели, но и категориальные значения (object), такие как "N", "n", "No tv". Для анализа переменной необходимо изменить другие значения помимо числовых на "0". Далее после обработки датасета выведите тип переменной.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [61]:
df['Time spent on TV'].unique()

array(['1', '0', '3', '0.5', 'n', '2', '4.5', '1.5', 'N', '4', '0.3', '5',
       'No tv', '0.1', '0.25', '6', '0.6', '7', '8', '15', ' ', '0.75',
       '2.5', '3.5', '0.4'], dtype=object)

In [62]:
df['Time spent on TV'] = df['Time spent on TV'].replace(to_replace=['n', 'N', 'No tv', ' '], value=0).astype(float)
print(df['Time spent on TV'].dtype)

float64


Взаимосвязано ли количество сна и время, проведенное в социальных сетях?

Для ответа на данный вопрос мы будем работать с двумя переменными 'Time spent on sleep' и 'Time spent on social media'. Мы хотим выяснить связано ли недосыпание учащихся с временем, проведенным ими в социальных сетях.

В начале перекодируйте переменные на бинарные категории: 'normal' и 'not normal'. Нормальное количество сна для людей – больше 7 часов. Нормальное время, проведенное в сети – это меньше 2 часов в день. Перекодируйте переменные, а затем воспользуйтесь результатом Хи-теста и выведите результат Хи-квадрата.

Одним из способом решения задачи являются следующие шаги:

    Воспользоваться циклом для создания новой переменной (название которой может быть 'Sleep') по переменной['Time spent on sleep'] , где необходимо прописать следующие условия. Если ['Time spent on sleep'] меньше 9, но при этом больше 6.9, то это нормальное количество сна ('normal'), в другом случае значение переменной будет 'not normal'.
    Воспользоваться циклом для создания еще одной новой переменной (название которой может быть 'Media') по переменной['Time spent on social media'] , где необходимо прописать следующие условия. Если ['Time spent on social media'] меньше 2, то это нормальное количество времяпровождения в социальных сетей ('normal'), в другом случае значение переменной будет 'not normal'.
    Воспользуйтесь функцией 'pd.crosstab' для подготовки данных к выдачи Хи-квадрата. crosstab() — метод, который помогает создать кросс-таблицу, показывает распределение данных по двум переменным. То есть вам необходимо применить метод  crosstab для двух, созданных вами, переменных (в качестве примера - это переменные 'Sleep' и 'Media').
    Воспользуйтесь функцией 'chi2_contingency' из пакета 'stats' и выведите значение Хи-квадрата.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [63]:
mask = (df['Time spent on sleep'] < 7) | (df['Time spent on sleep'] > 8.9)

df['Time spent on sleep'] = df['Time spent on sleep'].replace(to_replace=list(df[mask]['Time spent on sleep']), value= 'not normal')
df['Time spent on sleep'] = df['Time spent on sleep'].replace(to_replace=list(df[df['Time spent on sleep'] != 'not normal']['Time spent on sleep']), value= 'normal')

In [64]:
df['Time spent on sleep']

0           normal
1       not normal
2       not normal
3       not normal
4           normal
           ...    
1177        normal
1178    not normal
1179        normal
1180        normal
1181        normal
Name: Time spent on sleep, Length: 1182, dtype: object

In [65]:
mask2 = df['Time spent on social media'] < 2
df['Time spent on social media'] = df['Time spent on social media'].replace(to_replace=list(df[mask2]['Time spent on social media']), value= 'normal')

df['Time spent on social media'] = df['Time spent on social media'].replace(to_replace=list(df[df['Time spent on social media'] != 'normal']['Time spent on social media']), \
                                         value= 'not normal')

df['Time spent on social media']

0       not normal
1       not normal
2       not normal
3       not normal
4       not normal
           ...    
1177        normal
1178        normal
1179    not normal
1180        normal
1181        normal
Name: Time spent on social media, Length: 1182, dtype: object

In [66]:
df['Time spent on sleep'] = df['Time spent on sleep'].astype('category')
df['Time spent on social media'] = df['Time spent on social media'].astype('category')

In [67]:
df.dtypes

ID                                                                                      object
Region of residence                                                                     object
Age of Subject                                                                           int64
Time spent on Online Class                                                             float64
Rating of Online Class experience                                                       object
Medium for online class                                                                 object
Time spent on self study                                                               float64
Time spent on fitness                                                                  float64
Time spent on sleep                                                                   category
Time spent on social media                                                            category
Prefered social media platform                    

In [69]:
from scipy.stats import chi2_contingency

In [75]:
cross_table = pd.crosstab(df['Time spent on social media'], df['Time spent on sleep'])

In [76]:
cross_table

Time spent on sleep         normal  not normal
Time spent on social media                    
normal                         239         202
not normal                     414         327

In [79]:
results = chi2_contingency(cross_table)
results

Chi2ContingencyResult(statistic=0.24977164627176776, pvalue=0.6172359142282844, dof=1, expected_freq=array([[243.6319797, 197.3680203],
       [409.3680203, 331.6319797]]))

In [81]:
print(results.statistic)

0.24977164627176776


Взаимосвязано ли количество сна и время, проведенное на в социальных сетях? 2.0

Мы хотим выяснить, связано ли недосыпание учащихся с временем, проведенным ими в социальных сетях.

В начале перекодируйте переменные на бинарные категории: 'normal' и 'not normal'. Нормальное количество сна для людей – больше 7 часов. Нормальное время, проведенное в сети –это меньше 2 часов в день. Перекодируйте переменные, а затем воспользуйтесь результатом Хи-теста и выведите результат Хи-квадрата.

Напишите аналогичный код из предыдущей задачи. (Рассчитайте Хи-квадрат для переменных 'Time spent on sleep' и 'Time spent on social media') и выведите результат p-value.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [84]:
print(results.pvalue)

0.6172359142282844


Замена значений переменных с одних на другие

Для удобной работы категориальные переменные перекодируют в числовые. В этой задаче мы будем работать с бинарной переменной 'Health issue during lockdown'.

Перекодируйте значения переменной "YES" и "NO", на "1" и "0", соответственно.

В выдаче выведите количество значений по данной переменной.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [92]:
df['Health issue during lockdown'] = df['Health issue during lockdown'].replace(['YES', 'NO'], [1, 0])
print(df['Health issue during lockdown'].value_counts())

Health issue during lockdown
0    1021
1     161
Name: count, dtype: int64


Сколько читают учащиеся, справляясь со стрессом?

Чтобы понять находятся ли опеределенное слово, символ или другой показатель в значениях переменных, можно воспользоваться функцией contains(). Подробно об этой функции можно прочесть по ссылке https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.contains.html. В этой задаче вы будете работать с переменной 'Stress busters'. Выведите количество учащихся, которые используют книги в качестве профилактики борьбы со стрессом (встречаемость 'book' в значениях переменной).

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [96]:
print(sum(df['Stress busters'].str.contains('book')))

85


Самая популярная платформа среди социльных сетей

Определите наиболее популярную медиаплатформу в социальных сетях. (переменная 'Prefered social media platform')

Отфильтруйте пользователей, которые используют эту платформу для социальной сети и вычислите, в среднем сколько времени в день проводят в социальных медиа (переменная 'Time spent on social media') пользователи, которые предпочитают популярную социальную платформу. Результат округлите до сотых.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [124]:
most_pop = df['Prefered social media platform'].value_counts().idxmax()
mask = df['Prefered social media platform'] == most_pop

x = df[mask]['Time spent on social media'].mean()

print(f'{x:.2f}')

2.91


Кто проводит больше времени в социальных медиа?

В предыдущей задаче мы выяснили, что пользователи, которые предпочитают Инстаграм, используют социальные медиа в среднем около 2.91 часа.

Чтобы понять насколько пользователи Инстаграма отличаются от пользователей, отдающих предпочтения другой социальной сети, необходимо сделать подобную аналитику.

Определите платформу, в которой пользователи проводят больше всего времени в среднем.

Результат выведите через пробел. В начале укажите название предпочитаемой социальной сети, а затем среднее количество времени, затрачиваемое на пользование социальными сетями через пробел. Результат округлите до сотых.

Датасет доступен под названием 'COVID-19 Survey Student Responses.csv'

In [125]:
df.groupby(['Prefered social media platform'])['Time spent on social media'].mean()

Prefered social media platform
Elyment       1.000000
Facebook      2.778846
Instagram     2.912216
Linkedin      2.053279
None          0.000000
Omegle        2.000000
Quora         1.000000
Reddit        1.600000
Snapchat      3.750000
Talklife     10.000000
Telegram      1.333333
Twitter       2.982143
WhatsApp      1.000000
Whatsapp      1.935119
Youtube       2.234713
Name: Time spent on social media, dtype: float64

In [133]:
name = df.groupby(['Prefered social media platform'])['Time spent on social media'].mean().idxmax()
value = df.groupby(['Prefered social media platform'])['Time spent on social media'].mean().max()

print(f'{name} {value:.1f}')

Talklife 10.0
